src/pudl/transform/ferc1.py

```diff
@@ -6745,31 +6752,80 @@ def make_calculation_dimensions_explicit(
-        observed_dim = (
-            table_dimensions_ferc1[["table_name", "xbrl_factoid", dim_col]]
-            .drop_duplicates()  # bc there are dupes after we removed the other dim cols
-            .rename(
-                columns={
-                    "xbrl_factoid": "xbrl_factoid_calc",
-                    "table_name": "table_name_calc",
-                }
-            )
-        )
+        observed_dim = table_dimensions_ferc1[
+            on_cols + [dim_col]
+        ].drop_duplicates()  # bc there are dupes after we removed the other dim cols
         null_dim_mask = calc_comps_w_dims[dim_col].isnull()
         null_dim = calc_comps_w_dims[null_dim_mask].drop(columns=[dim_col])
         calc_comps_w_implied_dims = pd.merge(
             null_dim,
             observed_dim,
-            on=["table_name_calc", "xbrl_factoid_calc"],
+            on=on_cols,
             how="left",
         )
         calc_comps_w_explicit_dims = calc_comps_w_dims[~null_dim_mask]
         calc_comps_w_dims = pd.concat(
             [calc_comps_w_implied_dims, calc_comps_w_explicit_dims]
         )
     return calc_comps_w_dims
+
+
+def add_parent_dimensions(calc_comps, dimensions):
+    """Add in the dimensions of the parents in the calculation components.
+
+    First, we treat the totals in these dimension columns by assuming that any total we
+    observe is the sum of all non-total dimensions. We implement this by creating new
+    calculation component records which have a parent of dimension total and calculation
+    components of the non-total dimensinos.
+
+    Then, we assume that every parent factoid should have the same dimensions as its
+    calculation component dimensions.
+
+    Args:
+        calc_comps: a table of calculation component records which have had some manual
+            calculation fixes applied.
+        dimensions: list of dimension columns to check.
+    """
+    calc_comp_idx = [
+        "table_name_parent",
+        "xbrl_factoid_parent",
+        "table_name",
+        "xbrl_factoid",
+    ]
+    # for each dimension col, add a _parent column while
+    for dim in dimensions:
+        total_mask = calc_comps[dim] == "total"
+        total_w_subdim_components = (
+            pd.merge(
+                # the totals will become the parent record
+                calc_comps.loc[total_mask, calc_comp_idx + dimensions],
+                # the sub-components will become the calc component records
+                calc_comps[~total_mask],
+                on=calc_comp_idx,
+                how="left",
+                validate="1:m",
+                suffixes=("_parent", ""),
+            )
+            # these new total -> sub-dimension records should have the same
+            # parent and component fact/table so overwrite the og
+            .assign(
+                table_name_parent=lambda x: x.table_name,
+                xbrl_factoid_parent=lambda x: x.xbrl_factoid,
+            )
```

Comment:
What's the scenario where this isn't already the case?

Reply:
this whole step 1 in the treatment is generating *fully new* calculation records which specificly link the totals to their sub-components. so the parent values of the table_name/xbrl_factoid of these new records need to be the same as their child/components.

there are already valid calculation component records with total as a child dimension value that is say a part of a different calculation. Those records are not being overwritten. these records in `total_w_subdim_components` are just new and different records which communicate this total -> non-total calculation that was not previously encapsulated within the calculation components.